File: backend-java/src/test/java/com/skewwhiffy/auraltester/controller/KeySignatureControllerTest.java
package com.skewwhiffy.auraltester.controller;

import com.skewwhiffy.auraltester.notation.factory.InternalNotationFactory;
import com.skewwhiffy.auraltester.notation.model.clef.Clef;
import com.skewwhiffy.auraltester.notation.model.key.Key;
import com.skewwhiffy.auraltester.notation.model.note.AbsoluteNote;
import com.skewwhiffy.auraltester.service.AbcService;
import com.skewwhiffy.auraltester.test.util.TestData;
import lombok.val;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class KeySignatureControllerTest {
    @Mock
    private InternalNotationFactory internalNotationFactory;

    @Mock
    private AbcService abcService;

    @InjectMocks
    private KeySignatureController keySignatureController;
    private Clef clef;
    private String clefString;
    private String keySignature;
    private AbsoluteNote keySignatureNote;
    private String abc;

    @BeforeEach
    void setUp() {
        clef = TestData.random().clef();
        clefString = TestData.random().string();
        keySignature = TestData.random().string();
        keySignatureNote = TestData.random().absoluteNote();
        abc = TestData.random().string();
    }

    @Test
    void respondsCorrectly() {
        when(internalNotationFactory.clef(clefString)).thenReturn(clef);
        when(internalNotationFactory.getNote(keySignature)).thenReturn(keySignatureNote);
        when(abcService.getAbc(clef, Key.major(keySignatureNote.note())))
                .thenReturn(() -> abc);

        val actual = keySignatureController.get(clefString, Optional.of(keySignature));

        assertThat(actual.abc()).isEqualTo(abc);
    }
}

File: backend.kotlin/src/test/kotlin/com/skewwhiffy/controller/KeySignatureControllerTest.kt
package com.skewwhiffy.controller

import com.skewwhiffy.notation.factory.InternalNotationFactory
import com.skewwhiffy.notation.model.abc.AbcProvider
import com.skewwhiffy.notation.model.clef.Clef
import com.skewwhiffy.notation.model.key.Key
import com.skewwhiffy.notation.model.note.AbsoluteNote
import com.skewwhiffy.service.AbcService
import com.skewwhiffy.test.util.TestData
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.junit.jupiter.MockitoExtension

@ExtendWith(MockitoExtension::class)
class KeySignatureControllerTest {
  @Mock
  private lateinit var internalNotationFactory: InternalNotationFactory

  @Mock
  private lateinit var abcService: AbcService

  @InjectMocks
  private lateinit var keySignatureController: KeySignatureController
  private lateinit var clef: Clef
  private lateinit var clefString: String
  private lateinit var keySignature: String
  private lateinit var keySignatureNote: AbsoluteNote
  private lateinit var abc: String

  @BeforeEach
  fun `set up`() {
    clef = TestData.random.clef
    clefString = TestData.random.string
    keySignature = TestData.random.string
    keySignatureNote = TestData.random.absoluteNote
    abc = TestData.random.string
  }

  @Test
  fun `responds correctly`() {
    `when`(internalNotationFactory.clef(clefString)).thenReturn(clef)
    `when`(internalNotationFactory.getNote(keySignature)).thenReturn(keySignatureNote)
    `when`(abcService.getAbc(clef, Key(keySignatureNote.note))).thenReturn(abc.let {
      object : AbcProvider {
        override val abc: String = it
      }
    })

    val actual = keySignatureController.get(clefString, keySignature)

    assertThat(actual.abc).isEqualTo(abc)
  }
}